# Introduction {#asm-intro}
```{r mrmt-intro-reset, include=FALSE}
mrmt$set_this_rmd_file(ps_this_rmd_file = ifelse(rstudioapi::isAvailable(), 
                                                 rstudioapi::getSourceEditorContext()$path, 
                                                 rprojroot::thisfile()))
```
According to Wikipedia `r mrmt$add("Wikipedia2019")`, the term `Big Data` has been used since the 1990s. Some credit was given to John Mashey `r mrmt$add("Mashey1998")` for popularizing the term. Nowadays `Big Data` is used in connection with large companies, social media or governments which collect massive amounts of data and which are then using that data to draw certain conclusions about behaviors of customers, or followers or voters. The presidential election campaigns of Barack Obama were examples of how `Big Data` were used to access behaviors of voters `r mrmt$add("Issenberg2013")`. A different example is the use of `Big Data` in health care. An overview of the use of `Big Data` in health care is given in `r mrmt$add("Adibuzzaman2017")`. The collected health data is most likely not only used by research but also by insurance companies. Therefore this type of using data to make certain predictions that influences our daily lives is becoming a general interest also outside of the area of animal breeding. 


## Traditional Animal Breeding {#asm-traditional-animal-breeding}
In animal breeding the statistical analyses that have emerged together with the advent of `Big Data` technologies have long been applied when breeding values are predicted for whole populations of animals using BLUP-based approaches. The process of breeding value prediction uses statistical methods to assess the genetic potential of breeding animals in a population based on a large collection of data that characterize each animal in the population. The data used to predict the breeding values are collected not for having predicted breeding values in the first place, but the data are collected in the first place for quality control or management purposes. The prediction of breeding values can be viewed as a side product. In the area of cattle breeding, data collection consists of rather complex flows of information. The flow of information is shown in Figure \@ref(fig:datacollectionflow).

```{r datacollectionflow, echo=FALSE, hook_convert_odg=TRUE, fig_path="odg", fig.cap="Data Flow in an Animal Breeding Program"}
#rmddochelper::use_odg_graphic(ps_path = "odg/datacollectionflow.odg")
knitr::include_graphics(path = "odg/datacollectionflow.png")
```
 
 
## Genomic Selection {#asm-genomic-selection} 
The data flow shown in Figure \@ref(fig:datacollectionflow) contains the traditional evaluation of data to result in predicted breeding values. But it is missing the newest development in the breeding industry. This development is known as `Genomic Selection` (GS). GS was introduced by the work of [@Meuwissen2001]. The methods presented by [@Meuwissen2001] were only introduced into practical breeding programs when `r mrmt$add("Schaeffer2006")` showed the tremendous potential of saving costs for breeding programs. The use of __genomic__ information for the assessment of the genetic potential of all breeding animals represents the core of the evaluation approach presented by [@Meuwissen2001]. The term `genomic` is used because genetic markers which are evenly spaced over the complete genome are used as information source. Single Nucleotide Polymorphisms (SNP) are the most widely used marker model nowadays. SNPs are single positions in the genome that occur in different variants in the whole population. A description on how to identify SNPs in a population is given in `r mrmt$add("Czech2018")`. Potential use cases of SNPs are outlined by `r mrmt$add("SeidelJr.2010")` and `r mrmt$add("Pant2012")`. The genetic configuration of an SNP in a given population is shown in Figure \@ref(fig:snpgeneticconfiguration).

```{r snpgeneticconfiguration, echo=FALSE, hook_convert_odg=TRUE, fig_path="odg", fig.cap="Genetic Configuration of a Single Nucleotide Polymorphism (SNP)"}
#rmddochelper::use_odg_graphic(ps_path = "odg/snpgeneticconfiguration.odg")
knitr::include_graphics(path = "odg/snpgeneticconfiguration.png")
```
 
These SNPs can occur anywhere in the genome which means they can be observed in coding regions, in non-coding regions as well as in regulatory regions. In genomic selection, we are working with a large set of SNPs that are distributed over the complete genome. Hence some of the SNPs will be located close to genetic positions that are important for the expression of quantitative traits of interest. Such genetic positions which are related to quantitative traits are often called `Quantitative Trait Loci` (QTL). QTL themselves are difficult to detect and their inheritance is often manifested in complex modes. But due to the likely occurrence of several SNPs in the close proximity of a QTL, the inheritance of QTL alleles and of surrounding SNP alleles will not be independent due to linkage between SNPs and QTL. Such a linkage scenario between two SNPs flanking a QTL is shown in Figure \@ref(fig:linkagesnpqtl).

```{r linkagesnpqtl, echo=FALSE, hook_convert_odg=TRUE, fig_path="odg", fig.cap="Two SNPs flanking a QTL"}
#rmddochelper::use_odg_graphic(ps_path = "odg/linkagesnpqtl.odg")
knitr::include_graphics(path = "odg/linkagesnpqtl.png")
```

Although the QTL is likely to span a range of many positions on the chromosome, we can still assume the QTL to be bi-allelic with alleles $Q_1$ and $Q_2$. In theory, any SNP position can have four different alleles according to the four different bases. But when looking at different SNPs in real-world populations, most of them only show two alleles. Hence, for the two SNPs flanking the QTL shown in Figure \@ref(fig:linkagesnpqtl) they also have just two alleles $SNP1_1$, $SNP1_2$, $SNP2_1$ and $SNP2_2$. In genetics dependent mode of inheritance between neighboring loci (marker or QTL) is also called `linkage disequilibrium` (LD). This means that any joint allele frequency $Pr(SNP1_i, Q_j, SNP2_k)$ does not correspond to the product of the single allele frequencies of the two SNPs ($SNP1$ and $SNP2$) and the QTL. In a formula this can be written as

\begin{equation}
 Pr(SNP1_i, Q_j, SNP2_k) \ne Pr(SNP1_i) * Pr(Q_j) * Pr(SNP2_k)
 (\#eq:allelfreqsnpqtllinkage)
\end{equation}

Assuming that the QTL allele $Q_1$ is favorable for the expression of a given trait of interest and using the fact of LD as expressed in \@ref(eq:allelfreqsnpqtllinkage), the alleles of $SNP1$ and $SNP2$ which occur more frequently together with $Q_1$ are therefore also related to favorable expression levels of the trait of interest. In real breeding populations, the position of the QTL is unknown. But because we know the allelic configuration of a large number of SNP loci from many breeding animals, we can reliably relate SNP alleles and favorable expression levels of traits of interest. 


## Mono-Genic Model {#asm-mono-genic-model}
In quantitative genetics, the so-called mono-genic or single-locus model allows us to quantify the genetic potential of breeding animals in terms of breeding values. The standard reference in quantitative genetics in which also the mono-genic model is described is `r mrmt$add("Falconer1996")`. For a single locus, the breeding value depends on the allele frequencies at that locus and on the additive substitution effect which is often called $\alpha$. The mono-genic model for any given SNP locus in relation to the level of expression of a given trait of interest can be visualized in the following Figure \@ref(fig:monogenicsnpmodel).

```{r monogenicsnpmodel, echo=FALSE, hook_convert_odg=TRUE, fig_path="odg", fig.cap="Single-Locus Model for a Quantitative Trait"}
#rmddochelper::use_odg_graphic(ps_path = "odg/monogenicsnpmodel.odg")
knitr::include_graphics(path = "odg/monogenicsnpmodel.png")
```

In a real breeding population, we assume that the effect of all loci linked to the SNPs are purely additive. Hence any values for $d$ are all zero. As a consequence of that the breeding values at any given SNP position only depend on the allele frequencies of the SNP and the $a$ values at every SNP. The overall breeding value of a given animal is computed as the sum of all locus-specific breeding values. This overall breeding value is called __genomic breeding value__ (GBV). In order to get an estimate of such a GBV, we have to estimate all $a$ values at any SNP position. This estimation procedure can be done in one of the following two ways.

1. Two step approach
2. Single step approach


## Two Step Approach {#asm-two-step-approach}
In the two step approach the estimation of the $a$-values and the computation of the GBVs are done in two separate steps. For the estimation of the $a$ values for all SNPs, a reference population is defined. In dairy cattle breeding this reference population consists of all male breeding animals. In the recent past, the reference population has been augmented continuously with female animals. The animals in the reference population are all genotyped and they also all have phenotypic measurements^[Whenever phenotypic measurements are not available, traditionally predicted breeding values are transformed back into pseudo phenotypes which are then used to estimate $a$ values.] for the trait of interest. The estimation of the $a$ values amounts to estimating fixed effects in a linear model. We will see in later chapters of this course what methods are available to estimate these parameters. 

In the second step the estimates for all the $a$ values are used to compute the GBVs for all animals with genomic SNP information also for those outside of the reference population. The Figure \@ref(fig:twostepgs) tries to summarize the process graphically.

```{r twostepgs, echo=FALSE, hook_convert_odg=TRUE, fig_path="odg", fig.cap="Two Step Approach To Estimate Genomic Breeding Values"}
#rmddochelper::use_odg_graphic(ps_path = "odg/twostepgs.odg")
knitr::include_graphics(path = "odg/twostepgs.png")
```

The big advantage of the two step method is that once we have defined a good reference population which yields reliable estimates for the $a$ values, the computation of the GBV is a simple computation of just summing up the $a$ contributions with the correct sign determined by the SNP genotypes of the animals for which the GBVs should be determined. All animals with SNP genotypes can get GBV values. The difficult part in the two step approach is to define a reliable reference population and to determine good phenotypic measurements ($y$).


## Single Step Approach {#asm-single-step-approach}
The estimation of the $a$ values and the prediction of the genomic breeding values is done in one step using linear mixed effects models. In this single step evaluation animals with and without genomic information can get predicted genomic breeding values in a single analysis. One possibility to get to this predicted breeding values is via the use of `Genomic BLUP` (GBLUP). This will be the topic of a complete chapter in this course. The problem with the single step approach is to get an estimate of the covariance between animals with and without genomic information. This is a problem of ongoing research.


## Summary {#asm-summary}
The main difference between traditional predictions of breeding values using a BLUP animal model and the prediction of GBV is that the former uses the so called __infinitesimal__ model to assess the genetic potential and the latter uses sufficiently dense genomic information and uses a __polygenic__ model. This difference is illustrated in Figure \@ref(fig:infinitesimalvspolygenic).

```{r infinitesimalvspolygenic, echo=FALSE, hook_convert_odg=TRUE, fig_path="odg", fig.cap="Infinitesimal Versus Polygenic Model"}
#rmddochelper::use_odg_graphic(ps_path = "odg/infinitesimalvspolygenic.odg")
knitr::include_graphics(path = "odg/infinitesimalvspolygenic.png")
```

In the remaining chapters, different approaches for the prediction of GBVs are described. Chapter 2 gives a description of the multiple linear regression model and how it was tried to be used for GBV prediction by [@Meuwissen2001]. Chapter 3 introduces BLUP methodology in the context of predicting GBVs. In Chapter 4 the method called LASSO is introduced. Interestingly enough, this method is used very seldom in the area of animal breeding. Last but not least, chapter 5 makes an excursion into Bayesian estimation approaches. The Bayesian methods are important because they are used in practical breeding programs of Swiss Dairy cattle. 

