---
title: "Experiments With FLEM"
author: "Peter von Rohr"
date: 2019-02-28
output: html_notebook
---


## Disclaimer
Experiments with FLEM using the SNP data specified in the course notes.


## Data
The following dataset is given in the course notes. 


```{r dataflemsnpobs, echo=FALSE}
### # fix the number of animals
n_nr_animal <- 20
### # intercept
n_inter_cept <- 500
### # residual standard deviation
n_res_sd <- 12.13
### # vector of genotype value coefficients
vec_geno_value_coeff <- c(-1,0,1)
### # sample genotypes of unlinked SNP randomly
set.seed(436)
### # fix allele frequency of positive allele
n_prob_snps <- .45
### # genotypic values 
vec_geno_val <- c(27.2, 7.3)
n_nr_snp <- length(vec_geno_val)
### # put together the genotypes into a matrix
# mat_geno_snp <- matrix(c(sample(vec_geno_value_coeff, n_nr_animal, prob = c((1-n_prob_snps)^2, 
#                                                                    2*(1-n_prob_snps)*n_prob_snps, 
#                                                                    n_prob_snps^2), 
#                        replace = TRUE),
#                        sample(vec_geno_value_coeff, n_nr_animal, prob = c(n_prob_snps^2, 
#                                                                    2*(1-n_prob_snps)*n_prob_snps, 
#                                                                    (1-n_prob_snps)^2), 
#                        replace = TRUE)),
#                        nrow = n_nr_snp)
### # sample all genotypes at once
mat_geno_snp <- matrix(sample(vec_geno_value_coeff, n_nr_snp * n_nr_animal, prob = c((1-n_prob_snps)^2, 
                                                                   2*(1-n_prob_snps)*n_prob_snps, 
                                                                   n_prob_snps^2), 
                       replace = TRUE),
                       nrow = n_nr_snp)

mat_obs_y <- n_inter_cept + crossprod(mat_geno_snp, vec_geno_val) + rnorm(n = n_nr_animal, mean = 0, sd = n_res_sd)
### # mapping the -1,0,1 codes to genotypes
geno_code_map <- tibble::tibble(code = c(-1, 0, 1),
                            `SNP G` = c("$G_2G_2$", "$G_1G_2$", "$G_1G_1$"),
                            `SNP H` = c("$H_2H_2$", "$H_1H_2$", "$H_1H_1$"),
                            `Genotypic Value G` = c("$-a_G$", "$0$", "$a_G$"),
                            `Genotypic Value H` = c("$-a_H$", "$0$", "$a_H$"))
geno_code <- tibble::tibble(`Code G` = mat_geno_snp[1,],
                            `Code H` = mat_geno_snp[2,])

### # use dplyr to macht genotypes to codes
suppressPackageStartupMessages( require(dplyr) )
geno_code %>% 
  inner_join(geno_code_map, by = c("Code G" = "code")) %>%
  select(`SNP G`, `Genotypic Value G`) -> geno_snp_g
geno_code %>% 
  inner_join(geno_code_map, by = c("Code H" = "code")) %>%
  select(`SNP H`, `Genotypic Value H`) -> geno_snp_h
geno_snp_all <- bind_cols(geno_snp_g, geno_snp_h)
### # add the data
mat_obs_y_rounded <- round(mat_obs_y, digits = 0)
tbl_obs <- tibble::tibble(Observation = mat_obs_y_rounded[,1])
geno_snp_all %>% bind_cols(tbl_obs) -> tbl_all_data
### # add animal ids
tbl_all_data <- bind_cols(Animal = c(1:n_nr_animal),tbl_all_data)

# tbl_flemsnppbs <- tibble::tibble()
knitr::kable(tbl_all_data,
             booktabs = TRUE,
             longtable = TRUE,
             caption = "Animals With Two SNP Loci Affecting A Quantitative Trait",
             escape = FALSE)
```

## Specify The Model
The above dataset is to be analysed using the following model. For an observation $y_i$, we can write

$$y_i = \beta_0 + W_i \cdot a + \epsilon_i$$

where $\beta_0$ is the intercept, $a$ is the vector of additive SNP effects, $W_i$ is a row vector containing the genotype codes ($-1$, $0$ and $1$) and $\epsilon_i$ is the random error term.

The parameters $\beta_0$ and $a$ are combined into a single vector $b$ where

$$b = \left[\begin{array}{c} \beta_0 \\ a_G \\ a_H \end{array} \right]$$

and $X$ is the single design matrix. As a result, the model can be written in matrix-vector notation as

$$y = X \cdot b + \epsilon$$.

The design matrix $X$ is given by

```{r}
(mat_X_flem <- cbind(rep(1,n_nr_animal), t(mat_geno_snp)))
```

In order to see who the solutions of the flem can be determined, we check on the rank of $X$

```{r}
Matrix::rankMatrix(mat_X_flem)
```


Hence the solution can still be computed as

```{r}
n_bhat_flem <- crossprod((solve(crossprod(mat_X_flem))),crossprod(mat_X_flem,mat_obs_y_rounded));n_bhat_flem
```

The same with `lm()`


```{r}
tbl_geno_code_data <- dplyr::bind_cols(geno_code, tbl_obs)
lm_geno_code <- lm(Observation ~ `Code G` + `Code H`, data = tbl_geno_code_data)
summary(lm_geno_code)
```


## More SNP Loci
In reality the number of SNP loci is much larger than what we have used so far. For our experiment we increase the number of loci and generate a new genotype dataset. 

```{r}
### # total number of SNP
n_nr_snp_large <- 500
### # number of snp with non-zero effect
n_nr_nz_snp <- 10
### # vector of genotypic effects
vec_geno_val_large <- rep(0, n_nr_snp_large)
### # genotypic effect sampled from normal distribution
n_geno_mean <- 19.3
n_geno_sd <- 7.13
vec_geno_val_large[1:n_nr_nz_snp] <- rnorm(n = n_nr_nz_snp, mean = n_geno_mean, sd = n_geno_sd)
### # sample the genotype matrix
mat_geno_large <- matrix(sample(vec_geno_value_coeff, n_nr_snp_large * n_nr_animal, prob = c((1-n_prob_snps)^2, 
                                                                   2*(1-n_prob_snps)*n_prob_snps, 
                                                                   n_prob_snps^2), 
                       replace = TRUE),
                       nrow = n_nr_snp_large)
### # sample the observations, only the genetic part changes compared to the small example
mat_obs_y_large <- n_inter_cept + crossprod(mat_geno_large, vec_geno_val_large) + rnorm(n = n_nr_animal, mean = 0, sd = n_res_sd)
### # combine everything into a tibble
tbl_geno_large <- tibble::as.tibble(t(mat_geno_large))
colnames(tbl_geno_large) <- sapply(1:n_nr_snp_large, function(x) paste0("SNP", x), USE.NAMES = FALSE)
tbl_data_large <- bind_cols(tibble::tibble(Animal = 1:n_nr_animal), 
                            tibble::tibble(Observation = round(mat_obs_y_large[,1], digits = 0)), 
                            tbl_geno_large)

```

In this second example, we consider to the total number of loci to be `r n_nr_snp_large`. But for the genotypic values, we assume that only the first `r n_nr_nz_snp` have an effect on the observation. 


## Rank of Large Design Matrix
Extend design matrix of large dataset by intercept and compute rank

```{r}
mat_X_flem_large <- cbind(rep(1, n_nr_animal), t(mat_geno_large))
dim(mat_X_flem_large)
```

The rank

```{r}
Matrix::rankMatrix(mat_X_flem_large)
```

What is the result of $X^TX$ and `solve()` 

```{r}
mat_xtx_large <- crossprod(mat_X_flem_large)
solve(mat_xtx_large)
```

The dimension of $X^TX$ is 

```{r}
dim(mat_xtx_large)
```

The condition number is coming out of the eigenvalues

```{r}
vec_abs_ev <- abs(eigen(mat_xtx_large, only.values = TRUE)$values)
min(vec_abs_ev)/max(vec_abs_ev)
```


## Solutions To Normal Equations
### Pre-requisites
__Definition__: A general inverse $G$ of any matrix $A$ is defined as one of the matrices that satisfies $AGA = A$.

__Lemma 3__: $PX^TX = QX^TX$ implies $PX^T = QX^T$. Based on the fact that, if $X^TX = 0$ implying that $X=0$, we can apply the same fact with $(PX^TX - QX^TX)(P-Q)^T = (PX^T - QX^T)(PX^T - QX^T)^T = 0$ which is true because we started with $PX^TX = QX^TX$, hence the first factor is $0$. Further it can only be true, if $PX^T - QX^T = 0$ and therefore $PX^T = QX^T$.

### Material of Section 5 in Searle 1971
Independently of the fact that our design matrix $X$ for the large system is column rank deficient, the least squares normal equation still hold. Hence, we can still write

$$ X^TX \hat{b} = X^T y$$

But because $X^TX$ is singular, it cannot be inverted and the unique solution that was known from the regression model can no longer be obtained.

In what follows, we are replicating some of the material of section 5 in chapter 1 of Searle 1971. Special attention is given to properties of generalized inverses of symmetric matrices like $X^TX$. Four useful properties of such generalized inverses are given in Theorem 7. 

__Theorem 7__. When $G$ is a generlized invers of $X^TX$ then

(i) $G^T$ is also a generlized inverse of $X^TX$
(ii) $XGX^TX = X$, hence $GX^T$ is a generalized inverse of $X$
(iii) $XGX^T$ is invariant to $G$
(iv) $XGX^T$ is symmetric, whether $G$ is or not

__Proof__. By definition $G$ satisfies $X^TXGX^TX = X^TX$

(i) Transposing the relationship in the definition yields $X^TXG^TX^TX = X^TX$ which establishes property (i)
(ii) Follows from Lemma 3. Transpose both sides of the statement gives $X^TP^TX = X^TQ^TX$ and hence $P^TX = Q^TX$. For our case we have $P^T = XGX^T$ and $Q = I$


## Estimable Functions
In section 4 of chapter 5 (p 180) the concept of estimable functions is explained.


















