---
title: "Experiments With FLEM"
author: "Peter von Rohr"
date: 2019-02-28
output: html_notebook
---


## Disclaimer
Experiments with FLEM using the SNP data specified in the course notes.


## Data
The following dataset is given in the course notes. 


```{r dataflemsnpobs, echo=FALSE}
### # fix the number of animals
n_nr_animal <- 20
### # intercept
n_inter_cept <- 500
### # residual standard deviation
n_res_sd <- 12.13
### # vector of genotype value coefficients
vec_geno_value_coeff <- c(-1,0,1)
### # sample genotypes of unlinked SNP randomly
set.seed(436)
### # fix allele frequency of positive allele
n_prob_snps <- .45
### # genotypic values 
vec_geno_val <- c(27.2, 7.3)
n_nr_snp <- length(vec_geno_val)
### # put together the genotypes into a matrix
# mat_geno_snp <- matrix(c(sample(vec_geno_value_coeff, n_nr_animal, prob = c((1-n_prob_snps)^2, 
#                                                                    2*(1-n_prob_snps)*n_prob_snps, 
#                                                                    n_prob_snps^2), 
#                        replace = TRUE),
#                        sample(vec_geno_value_coeff, n_nr_animal, prob = c(n_prob_snps^2, 
#                                                                    2*(1-n_prob_snps)*n_prob_snps, 
#                                                                    (1-n_prob_snps)^2), 
#                        replace = TRUE)),
#                        nrow = n_nr_snp)
### # sample all genotypes at once
mat_geno_snp <- matrix(sample(vec_geno_value_coeff, n_nr_snp * n_nr_animal, prob = c((1-n_prob_snps)^2, 
                                                                   2*(1-n_prob_snps)*n_prob_snps, 
                                                                   n_prob_snps^2), 
                       replace = TRUE),
                       nrow = n_nr_snp)

mat_obs_y <- n_inter_cept + crossprod(mat_geno_snp, vec_geno_val) + rnorm(n = n_nr_animal, mean = 0, sd = n_res_sd)
### # mapping the -1,0,1 codes to genotypes
geno_code_map <- tibble::tibble(code = c(-1, 0, 1),
                            `SNP G` = c("$G_2G_2$", "$G_1G_2$", "$G_1G_1$"),
                            `SNP H` = c("$H_2H_2$", "$H_1H_2$", "$H_1H_1$"),
                            `Genotypic Value G` = c("$-a_G$", "$0$", "$a_G$"),
                            `Genotypic Value H` = c("$-a_H$", "$0$", "$a_H$"))
geno_code <- tibble::tibble(`Code G` = mat_geno_snp[1,],
                            `Code H` = mat_geno_snp[2,])

### # use dplyr to macht genotypes to codes
suppressPackageStartupMessages( require(dplyr) )
geno_code %>% 
  inner_join(geno_code_map, by = c("Code G" = "code")) %>%
  select(`SNP G`, `Genotypic Value G`) -> geno_snp_g
geno_code %>% 
  inner_join(geno_code_map, by = c("Code H" = "code")) %>%
  select(`SNP H`, `Genotypic Value H`) -> geno_snp_h
geno_snp_all <- bind_cols(geno_snp_g, geno_snp_h)
### # add the data
mat_obs_y_rounded <- round(mat_obs_y, digits = 0)
tbl_obs <- tibble::tibble(Observation = mat_obs_y_rounded[,1])
geno_snp_all %>% bind_cols(tbl_obs) -> tbl_all_data
### # add animal ids
tbl_all_data <- bind_cols(Animal = c(1:n_nr_animal),tbl_all_data)

# tbl_flemsnppbs <- tibble::tibble()
knitr::kable(tbl_all_data,
             booktabs = TRUE,
             longtable = TRUE,
             caption = "Animals With Two SNP Loci Affecting A Quantitative Trait",
             escape = FALSE)
```

## Specify The Model
The above dataset is to be analysed using the following model. For an observation $y_i$, we can write

$$y_i = \beta_0 + W_i \cdot a + \epsilon_i$$

where $\beta_0$ is the intercept, $a$ is the vector of additive SNP effects, $W_i$ is a row vector containing the genotype codes ($-1$, $0$ and $1$) and $\epsilon_i$ is the random error term.

The parameters $\beta_0$ and $a$ are combined into a single vector $b$ where

$$b = \left[\begin{array}{c} \beta_0 \\ a_G \\ a_H \end{array} \right]$$

and $X$ is the single design matrix. As a result, the model can be written in matrix-vector notation as

$$y = X \cdot b + \epsilon$$.

The design matrix $X$ is given by

```{r}
(mat_X_flem <- cbind(rep(1,n_nr_animal), t(mat_geno_snp)))
```

In order to see who the solutions of the flem can be determined, we check on the rank of $X$

```{r}
Matrix::rankMatrix(mat_X_flem)
```


Hence the solution can still be computed as

```{r}
n_bhat_flem <- crossprod((solve(crossprod(mat_X_flem))),crossprod(mat_X_flem,mat_obs_y_rounded));n_bhat_flem
```

The same with `lm()`


```{r}
tbl_geno_code_data <- dplyr::bind_cols(geno_code, tbl_obs)
lm_geno_code <- lm(Observation ~ `Code G` + `Code H`, data = tbl_geno_code_data)
summary(lm_geno_code)
```


## More SNP Loci
In reality the number of SNP loci is much larger than what we have used so far. For our experiment we increase the number of loci and generate a new genotype dataset. 

```{r}
### # total number of SNP
n_nr_snp_large <- 500
### # number of snp with non-zero effect
n_nr_nz_snp <- 10
### # vector of genotypic effects
vec_geno_val_large <- rep(0, n_nr_snp_large)
### # genotypic effect sampled from normal distribution
n_geno_mean <- 19.3
n_geno_sd <- 7.13
vec_geno_val_large[1:n_nr_nz_snp] <- rnorm(n = n_nr_nz_snp, mean = n_geno_mean, sd = n_geno_sd)
### # sample the genotype matrix
mat_geno_large <- matrix(sample(vec_geno_value_coeff, n_nr_snp_large * n_nr_animal, prob = c((1-n_prob_snps)^2, 
                                                                   2*(1-n_prob_snps)*n_prob_snps, 
                                                                   n_prob_snps^2), 
                       replace = TRUE),
                       nrow = n_nr_snp_large)

```

For the genotypic values, we assume that only the first `r n_nr_nz_snp` have an effect on the observation. 






