---
title: Applied Statistical Methods - Exercise 3
author: Peter von Rohr
date: 2019-03-11
output:
  bookdown::pdf_document2:
    toc: false
    number_sections: false
    keep_tex: false
header-includes:
 \usepackage{longtable}
 \usepackage{float}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, results = 'asis', fig.pos = 'H')
knitr::knit_hooks$set(hook_convert_odg = rmddochelper::hook_convert_odg)
```


## Problem 1: Fixed Linear Effects Model {-}
We want to analyse a dataset with genetic information using a fixed linear effects model. The dataset is taken from the course notes and is shown in Table \@ref(tab:datatable). 

We assume that the SNP loci have a purely additive effect on the trait. That means for a SNP locus $L$ the absolute value of the genotypic value of the homozygous genotypes ($L_1L_1$ and $L_2L_2$) is taken to be $a_L$ and the genotypic value of the heterozygous genotype ($L_1L_2$) is taken to be $0$. The fixed linear effects model contains the observation in Table \@ref(tab:datatable) as the response variable, an intercept and the genotypic values of the the genotypes at the two SNP Loci $G$ and $H$ as predictor variables.

For the observation $y_i$ of animal $i$, we can specify the model as

$$y_i = \beta_0 + W_i \cdot a + \epsilon_i$$

where $\beta_0$ is the intercept, $a$ is the vector of additive SNP-effects, $W_i$ is a row vector denoting the SNP-Genotypes and $\epsilon_i$ is the random error term.

The data can be read from `https://charlotte-ngs.github.io/GELASMSS2019/ex/w04/ex03p01_data.csv`. It can be read using the following statement

```{r datainput, message=FALSE}
### # specify path to data file depending on online status
s_data_file <- "https://charlotte-ngs.github.io/GELASMSS2019/ex/w04/ex03p01_data.csv"
### # read the data into a tibble
tbl_geno_data <- readr::read_csv(file = s_data_file)
```

```{r datatable, echo=FALSE, results='asis'}
knitr::kable(tbl_geno_data,
             booktabs = TRUE, 
             longtable = TRUE, 
             caption = "Genotypic Data Used for Fitting a Fixed Linear Effect Model")
```

### Your Tasks {-}
* Specify the fixed linear effects model in matrix-vector notation by putting the information from the dataset into the model. Use the same parametrization as shown in the course notes where the intercept $\beta_0$ and the vector $a$ are combined into a single parameter vector $b$. The design matrix that links elements in $b$ to observations $y$ is then called $X$. 
* Use the function `Matrix::rankMatrix()` from the `Matrix` package on the matrix $X$ to find out the rank of the design matrix.
* Depending on the rank of $X$ compute an estimate for $b$, if the rank of the matrix is equal to the number of columns of matrix $X$, then the same forumla as was used in the regression model can be used
* Verify your results using the `lm()` function


### Hints {-}
* Read the data using the function `readr::read_csv()`



## Problem 2: Genomic Relationship Matrix {-}
From the given dataset that can be obtained from 

`https://charlotte-ngs.github.io/GELASMSS2019/ex/w04/ex03p02_data.csv`, 

compute the genomic relationship matrix $G$. The dataset is organised such that animals are in rows and SNPs are in columns.


### Hints {-}
* Read the data using the function `readr::read_csv()`
* Convert the input data with the function `as.matrix()` to a matrix
* Use the function `apply(mat, 2, mean)` to compute the columnwise mean of matrix `mat`
* Use the `matrix` function to construct the matrix $P$ from the vector of SNP allele frequencies




## Additional Problem {-}
Write a function in R that accepts a matrix of genotypes and that computes the genomic relationship matrix. Verify your results from Problem 2.


