---
title:  Applied Statistical Methods -- Excercise 1
author: Peter von Rohr
date: 2019-02-18
output: 
  bookdown::pdf_document2:
    toc: FALSE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, results = 'asis')
knitr::knit_hooks$set(hook_convert_odg = rmddochelper::hook_convert_odg)
```


## Problem 1: Breeding Values {-}
During the lecture the computation of the breeding values for a given genotype was shown for a completely additive locus which means the genotypic value $d$ of the heterozygous genotypes is $0$. In this exercise, we want to compute the general solution for the breeding values of all three genotypes under a monogenic model. We are given a single locus $G$ with two alleles $G_1$ and $G_2$ which are closely linked to a QTL for a trait of interest. We assume that the population is in Hardy-Weinberg equilibrium at the given locus $G$. The allele frequencies are 

```{r allelefreqtable}
tbl_allelefreq <- tibble::tibble(Allele = c("$G_1$", "$G_2$"),
                                 Frequency = c("$p$", "$q$"))
knitr::kable(tbl_allelefreq, 
             booktabs = TRUE,
             longtable = TRUE,
             escape = FALSE)
```

Allele $G_1$ is the one with a positive effect on the trait of interest. The genotypic values are given in the following table.

```{r genovalue}
tbl_genovalue <- tibble::tibble(Genotype = c("$G_1G_1$", "$G_1G_2$", "$G_2G_2$"),
                                Value    = c("$a$", "$d$", "$-a$"))
knitr::kable(tbl_genovalue,
             booktabs = TRUE,
             longtable = TRUE,
             escape = FALSE)
```


### Your Task {-}

* Compute the breeding values for all three genotypes $G_1G_1$, $G_1G_2$ and $G_2G_2$.
* Verify the results presented in the lecture by setting $d=0$ in the breeding values you computed before.


\clearpage
\pagebreak

## Problem 2: Linkage Between SNP and QTL {-}
In a population of breeding animals, we are given a trait of interest which is determined by a QTL $Q$  on chromosome $1$. QTL $Q$ is modeled as a bi-allelic QTL with alleles $Q_1$ and $Q_2$. Furthermore we have genotyped our population for two SNPs $R$ and $S$ with two alleles each. One of the SNPs is on chromosome $1$ and is closely linked to $Q$. The other SNP is on chromosome $2$ and is unlinked. Figure \@ref(fig:linkageqtlsnp) shows the situation in a diagram.

```{r linkageqtlsnp, echo=FALSE, hook_convert_odg=TRUE, fig_path="odg", fig.cap="Linkage Between an SNP and a QTL and an independent SNP on a different Chromosome"}
#rmddochelper::use_odg_graphic(ps_path = "odg/linkageqtlsnp.odg")
knitr::include_graphics(path = "odg/linkageqtlsnp.png")
```
 
Based on the following small dataset, determine which of the two SNPs $R$ and/or $S$ is linked to QTL $Q$.

```{r problem2data}
### # fix the number of animals
n_nr_animal <- 12
### # fix number of snp columns
n_nr_snp <- 2
### # residual standard deviation
n_res_sd <- 2.13
### # vector of genotype value coefficients
vec_geno_value_coeff <- c(-1,0,1)
### # sample genotypes of unlinked SNP randomly
set.seed(9876)
### # fix allele frequency of positive allele
n_prob_snps <- .5
### # genotypic values 
vec_geno_val <- c(0, 23.52)
### # put together the genotypes into a matrix
mat_geno_snp <- matrix(c(sample(vec_geno_value_coeff, n_nr_animal, replace = TRUE),
                         sample(vec_geno_value_coeff, n_nr_animal, prob = c((1-n_prob_snps)^2, 2*(1-n_prob_snps)*n_prob_snps, n_prob_snps^2), replace = TRUE)),
                       nrow = n_nr_snp)
### # compute the observations
mat_obs_y <- crossprod(mat_geno_snp, vec_geno_val) + rnorm(n = n_nr_animal, mean = 0, sd = n_res_sd)
### # convert them to a tibble and round to two digits
tbl_obs <- tibble::tibble(Observation = round(mat_obs_y[,1], digits = 2))

### # create table with genotypes in string format which is done
### #  via a common mapping tibble
geno_code_map <- tibble::tibble(code = c(-1, 0, 1),
                            `SNP R` = c("$R_2R_2$", "$R_1R_2$", "$R_1R_1$"),
                            `SNP S` = c("$S_2S_2$", "$S_1S_2$", "$S_1S_1$"))
### # genotypes in -1, 0, 1 coding are collected in a tibble
geno_code <- tibble::tibble(`Code R` = mat_geno_snp[1,],
                            `Code S` = mat_geno_snp[2,])
### # map the coded genotypes to the string formats 
suppressPackageStartupMessages( require(dplyr) )
geno_code %>% 
  inner_join(geno_code_map, by = c("Code R" = "code")) %>%
  select(`SNP R`) -> geno_snp_r
geno_code %>% 
  inner_join(geno_code_map, by = c("Code S" = "code")) %>%
  select(`SNP S`) -> geno_snp_s
### # bind the genotypes for the two SNPs together
geno_snp_r %>%
  bind_cols(geno_snp_s) -> tbl_all_geno
### # bind genotypes and obserations into one tibble
tbl_all_geno %>% bind_cols(tbl_obs) -> tbl_all_data
### # produce the table
knitr::kable(tbl_all_data, 
             booktabs = TRUE,
             longtable = FALSE,
             escape = FALSE,
             caption = "Dataset showing linkage between SNP and QTL") 
```

```{r}
### # Write the data to a file. To do that we use an ascii-based coding map
geno_code_map_ascii <- tibble::tibble(code = c(-1, 0, 1),
                            `SNP R` = c("R2R2", "R1R2", "R1R1"),
                            `SNP S` = c("S2S2", "S1S2", "S1S1"))
geno_code %>% 
  inner_join(geno_code_map_ascii, by = c("Code R" = "code")) %>%
  select(`SNP R`) -> geno_snp_r_ascii
geno_code %>% 
  inner_join(geno_code_map_ascii, by = c("Code S" = "code")) %>%
  select(`SNP S`) -> geno_snp_s_ascii
### # bind the genotypes for the two SNPs together
geno_snp_r_ascii %>%
  bind_cols(geno_snp_s_ascii) -> tbl_all_geno_ascii
### # bind genotypes and obserations into one tibble
tbl_all_geno_ascii %>% bind_cols(tbl_obs) -> tbl_all_data_ascii
### # write the ascii-formatted data to a file
s_asm_w02_ex01_p02_genodatafile <- "asm_w02_ex01_p02_genodatafile.csv"
if (!file.exists(s_asm_w02_ex01_p02_genodatafile))
  readr::write_csv(tbl_all_data_ascii, path = s_asm_w02_ex01_p02_genodatafile)
```

From the above table it might be difficult to decide which SNP is linked to the QTL. Plotting the data may help. Showing the observations as a function of the genotypes leads to Figure \@ref(fig:problem2plot).

```{r problem2plot, fig.show='hold', out.width='50%', fig.cap="Observations Grouped by SNP Genotypes"}
suppressPackageStartupMessages( require(ggplot2) )
ggplot(data = tbl_all_data_ascii, aes(x = `SNP R`, y = Observation)) + 
  geom_point(color = 'blue')
ggplot(data = tbl_all_data_ascii, aes(x = `SNP S`, y = Observation)) + 
  geom_point(color = 'green') 
```



### Your Tasks {-}
* Determine which of the two SNPs $R$ or $S$ is closely linked to the QTL
* Estimate a value for $a$ obtained based on the data
* Try to fit a linear model through the genotypes that SNP which is linked to the QTL using the `lm()` function. The genotype data is available from

`https://charlotte-ngs.github.io/GELASMSS2019/ex/w02/asm_w02_ex01_p02_genodatafile.csv`


